rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+             Composite="cutlass.dense_bias") -> Tensor[(1, 10), float32] {
+      %0 = nn.dense(%FunctionVar_01, %FunctionVar_11, units=None, out_dtype="float32");
+      add(%0, %FunctionVar_21)
+    };
+    %1(%FunctionVar_0, %FunctionVar_1, %FunctionVar_2)
+  };
+  %20(%19, meta[relay.Constant][4] /*Tensor[(10, 256), float32]*/,
+      meta[relay.Constant][5] /*Tensor[(1, 10), float32]*/)
+}
+```
+
+Ideally this optimal partitioning would be understandable to the user, see Appendix G.
+
+## Design
+
+The implementation is mostly under `src/relay/collage/...` (namespace `tvm::relay::collage`), with just a few Python
+helper functions under `python/tvm/relay/collage`.
+
+If the `relay.collage.enable_collage` `PassConfig` attribute is true then a new `CollagePartitioner` pass is inserted
+before all other Relay passes. The result of the pass is:
+
+- All Relay sub-graphs in all global functions which are to be handed off to a BYOC toolchain are replaced by calls to
+  an inline `"Primitive"` function with `"Compiler"` and `"global_symbol"` attributes.
+- Relay operators, or groups of operators, which are to be translated to particular library or BYOC-supplied function
+  are replaced by calls to an inline `"Composite"` function. (This encoding is supported for both BYOC and external
+  libraries.)
+
+Note that no `"Primitive"` functions denoting TVM kernels are produced -- the existing `FuseOps` pass is still required.
+
+The `CollagePartitioner` pass has four phases:
+
+- **Phase 1**: The available `Target`s are scanned to build a list of rules describing how to find possible partitions (
+  see `PartitionSpec` and `PartitionRule` below). Depending on the `Target` the rules may incorporate entries from the
+  BYOC pattern table. (The remaining phases execute on each global function separately.)
+- **Phase 2**: A dataflow graph is constructed for the global function (which is just an `IndexedGraph<Expr>`). The
+  available rules from phase 1 are evaluated on the dataflow graph to yield a (possibly overlapping) set of candidate
+  partitions for each target (see `CandidatePartition` below). Each candidate efficiently describes a sub-graph of the
+  global function's body without the need to construct any new expressions (see `SubGraph` below).
+- **Phase 3**: A shortest path is found in the following (implicit and lazily constructed) search graph:
+    - Search Nodes: The set of dataflow nodes which have already been assigned to a candidate partition in all paths to
+      the node.
+    - Search Edge X->Y: A candidate partition can be applied to node X to give node Y. The candidate is disjoint from
+      all dataflow nodes already assigned in X. To avoid an unnecessary search space explosion the candidate must also
+      include the next yet-to-be-assigned dataflow node in X.
+    - Edge cost: The estimated latency of the candidate partition, plus a partition transition penalty. Note that though
```

Comment:
[Clarification] why is estimated latency associated with an edge ? (partition transition penalty make sense)

Reply:
I'm still following the paper where this is considered a 'kernel launch penalty', which is uniquely determined by the partition, which is captured in the search graph edges. I don't have any experience with this aspect, and don't yet know whether i) we'll need to measure these penalties, ii) their impact on optimal placements and iii) whether they should be determined by the pair of successive partitions rather than just one partition. So I'm leaving this a bit vague.